IdentityCore/src/cache/mac/MSIDMacKeychainTokenCache.m

```diff
@@ -870,6 +973,121 @@ - (BOOL)clearWithContext:(id<MSIDRequestContext>)context
+        }
+        
+        return self.accessForSharedBlob;
+    }
+    else
+    {
+        if (!self.accessForNonSharedBlob)
+        {
+            NSArray* trustedApps = [self createAccessForCurrentApp:context error:error];
+            self.accessForNonSharedBlob = [self accessCreateWithChangeACL:trustedApps
+                                                                  context:context
+                                                                    error:error];
+        }
+        
+        return self.accessForNonSharedBlob;
+    }
+}
+
+- (NSArray *)createAccessForCurrentApp:(id<MSIDRequestContext>)context error:(__unused NSError **)error
+{
+    SecTrustedApplicationRef trustedApplication = nil;
+    OSStatus status = SecTrustedApplicationCreateFromPath(nil, &trustedApplication);
+    if (status != errSecSuccess)
+    {
+        [self createError:@"Failed to create trusted application for current application path"
+                   domain:MSIDKeychainErrorDomain errorCode:status error:error context:context];
+        
+        return nil;
+    }
+    
+    NSArray *trustedApplications = @[(__bridge_transfer id)trustedApplication];
+    return trustedApplications;
+}
+
+- (id)accessCreateWithChangeACL:(NSArray<id> *)trustedApplications
+                         context:(id<MSIDRequestContext>)context
+                           error:(NSError **)error
+{
+    SecAccessRef access;
+    OSStatus status = SecAccessCreate((__bridge CFStringRef)s_defaultKeychainLabel, (__bridge CFArrayRef)trustedApplications, &access);
+    
+    if (status != errSecSuccess)
+    {
+        [self createError:@"Failed to create SecAccessRef object for blob."
+                   domain:MSIDKeychainErrorDomain errorCode:status error:error context:context];
+        return nil;
+    }
+    
+    if (![self accessSetACLTrustedApplications:access
+                           aclAuthorizationTag:kSecACLAuthorizationChangeACL
+                           trustedApplications:trustedApplications
+                                       context:context
+                                         error:error])
+    {
+        [self createError:@"Failed to update ACL with kSecACLAuthorizationChangeACL authorization tag and set trusted applications."
+                   domain:MSIDKeychainErrorDomain errorCode:status error:error context:context];
+        
+        CFRelease(access);
+        return nil;
+    }
+    
+    return CFBridgingRelease(access);
+}
+
+- (BOOL)accessSetACLTrustedApplications:(SecAccessRef)access
+                     aclAuthorizationTag:(CFStringRef)aclAuthorizationTag
+                     trustedApplications:(NSArray<id> *)trustedApplications
+                                 context:(id<MSIDRequestContext>)context
+                                   error:(NSError **)error
+{
+    OSStatus status;
+    NSArray *acls = (__bridge_transfer NSArray*)SecAccessCopyMatchingACLList(access, aclAuthorizationTag);
```

Comment:
can copy fail?

Reply:
This api does not take an error object. It is possible that this list might return zero items if we pass in an authorization tag that does not correspond to any acl. We are passing kSecAuthorizationChangeACL which is always present as one of the authorization tags so it is very unlikely that we get an empty list.